pco/src/float_quant_utils.rs

```diff
@@ -0,0 +1,75 @@
+use std::cmp;
+
+use crate::constants::Bitlen;
+use crate::data_types::{FloatLike, Latent};
+use crate::sampling;
+
+#[inline(never)]
+pub(crate) fn join_latents<F: FloatLike>(
+  k: Bitlen,
+  primary: &mut [F::L],
+  secondary: &[F::L],
+) -> () {
+  for (y_and_dst, &m) in primary.iter_mut().zip(secondary.iter()) {
+    debug_assert!(
+      (m >> k).is_zero(),
+      "Invalid input to FloatQuant: m must be a k-bit integer"
+    );
+    *y_and_dst = (*y_and_dst << k) + m;
```

Comment:
We wanted to change this to something that's monotonic in y, right?

Reply:
AIUI, it is.  Note that `y` is an `L`, as this function is only supposed to be called on the output of `split_latents`.  And in `split_latents`, there is a call to `to_latent_ordered` that makes the first output monotonic in the float representation of the input `F`.

Reply:
You're right. We should still solve the problem that `m` differs for perfectly quantized positive vs negative floats, though. Here are two proposals (I recommend we go with the first one):

1. We tweak flip the meaning of `m` when the float is negative. I think this would look like `let sign_cutoff = L::MID >> k; let ulps_away = if y >= cutoff {m} else {(1 << k) - 1 - m}; *dst = (y << k) + ulps_away`. This should have all the properties we want.
2. (the meaning of `y` and `m` is actually identical to 1.) We do away with the mandatory final step of `from_latent_ordered` and do `let quantums = if y >= L::MID {y - L::MID} else {L::MAX - y} ; let num = from_bits((mid_centered_quantums << k) + m)`. I think this is aesthetically better with slightly better performance, but less consistent with how I wrote the other modes, and we'd have to change some of the other code to avoid calling `from_latent_ordered`.

Reply:
Yeah, that sounds good.  Just implemented # 1 and pushed.

Fwiw, I'm generally of the view that any trade-offs made in the name of performance should be accompanied by some kind of repeatable profiling that shows the performance gain is real -- caching and register allocation (not to mention compiler optimizations) are too complicated to be guessable in most cases.  My impression is that you do a fair bit of this sort of thing already, so I'm probably not saying anything new there :slightly_smiling_face:

As for the aesthetics stuff, one strategy I sometimes use is to have two implementations of a tricky snippet of code: in the docs, an elegantly explainable, possibly-slower reference implementation; and in the code, a performant, possibly not-easy-to-explain implementation, accompanied by a unit test confirming that the two implementations are functionally identical.